File: backend/src/main/java/com/joying/chat/service/UnreadCountService.java
package com.joying.chat.service;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executor;
import java.util.concurrent.TimeUnit;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.stereotype.Service;

import com.joying.chat.domain.ChatRoomMember;
import com.joying.chat.repository.ChatMessageRepository;
import com.joying.chat.repository.ChatRoomMemberRepository;

/**
 * 안읽은 메시지 개수.
 *
 * <p>Redis를 먼저 보고, 없으면 저장소에서 세어 다시 넣는다. 메시지가 올 때마다 세면
 * 방 목록을 열 때마다 방 수만큼 집계 쿼리가 나간다.
 *
 * <p>Redis가 죽어도 메시지 저장을 되돌리지 않는다. 배지가 잠깐 틀린 것과 메시지가
 * 사라지는 것 중 뒤엣것이 훨씬 비싸다.
 */
@Service
public class UnreadCountService {

	private static final Logger log = LoggerFactory.getLogger(UnreadCountService.class);

	private static final String UNREAD_KEY_PREFIX = "unread:";
	private static final long TTL_DAYS = 7L;

	private final RedisTemplate<String, String> redis;
	private final ChatMessageRepository chatMessageRepository;
	private final ChatRoomMemberRepository chatRoomMemberRepository;
	private final Executor queryExecutor;

	public UnreadCountService(RedisTemplate<String, String> redis,
							  ChatMessageRepository chatMessageRepository,
							  ChatRoomMemberRepository chatRoomMemberRepository,
							  @Qualifier("chatQueryExecutor") Executor queryExecutor) {
		this.redis = redis;
		this.chatMessageRepository = chatMessageRepository;
		this.chatRoomMemberRepository = chatRoomMemberRepository;
		this.queryExecutor = queryExecutor;
	}

	private String getKey(Long chatRoomId, Long memberId) {
		return UNREAD_KEY_PREFIX + chatRoomId + ":" + memberId;
	}

	/**
	 * 메시지를 받은 쪽의 개수를 올린다.
	 */
	public void increment(Long chatRoomId, Long memberId) {
		try {
			String key = getKey(chatRoomId, memberId);
			redis.opsForValue().increment(key);
			redis.expire(key, TTL_DAYS, TimeUnit.DAYS);
		} catch (Exception e) {
			// 저장은 이미 끝났다. 여기서 터뜨려 그것까지 되돌리지 않는다.
			log.error("Redis increment 실패: {}", e.getMessage());
		}
	}

	/**
	 * 읽었으므로 0으로 되돌린다.
	 */
	public void reset(Long chatRoomId, Long memberId) {
		try {
			redis.delete(getKey(chatRoomId, memberId));
		} catch (Exception e) {
			log.error("Redis reset 실패: {}", e.getMessage());
		}
	}

	public long get(Long chatRoomId, Long memberId) {
		String cached = redis.opsForValue().get(getKey(chatRoomId, memberId));
		if (cached != null) {
			return Long.parseLong(cached);
		}
		return warmup(chatRoomId, memberId);
	}

	/**
	 * 여러 방을 한 번에 본다.
	 *
	 * <p>Redis는 한 번에 묶어 묻고, 없는 것만 저장소에서 센다. 방마다 저장소를 봐야
	 * 하므로 순차로 하면 방 수만큼 지연이 더해진다. 서로 의존하지 않으니 동시에 날린다.
	 */
	public Map<Long, Long> getBatch(List<Long> chatRoomIds, Long memberId) {
		if (chatRoomIds.isEmpty()) {
			return Map.of();
		}

		List<String> keys = chatRoomIds.stream().map(id -> getKey(id, memberId)).toList();

		List<String> cached;
		try {
			List<String> result = redis.opsForValue().multiGet(keys);
			cached = result == null ? List.of() : result;
		} catch (Exception e) {
			log.error("Redis multiGet 실패: {}", e.getMessage());
			cached = List.of();
		}

		Map<Long, Long> result = new HashMap<>();
		List<Long> missedIds = new ArrayList<>();

		for (int i = 0; i < chatRoomIds.size(); i++) {
			Long chatRoomId = chatRoomIds.get(i);
			Long count = i < cached.size() ? parseOrNull(cached.get(i)) : null;
			if (count != null) {
				result.put(chatRoomId, count);
			} else {
				missedIds.add(chatRoomId);
			}
		}

		log.debug("Redis 배치 조회: 총 {}개, 캐시 히트 {}개, 캐시 미스 {}개",
			chatRoomIds.size(), result.size(), missedIds.size());

		if (!missedIds.isEmpty()) {
			List<CompletableFuture<Map.Entry<Long, Long>>> futures = missedIds.stream()
				.map(chatRoomId -> CompletableFuture.supplyAsync(
					() -> Map.entry(chatRoomId, warmup(chatRoomId, memberId)), queryExecutor))
				.toList();

			CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();
			futures.forEach(future -> {
				Map.Entry<Long, Long> entry = future.join();
				result.put(entry.getKey(), entry.getValue());
			});
		}

		return result;
	}

	private Long parseOrNull(String value) {
		if (value == null) {
			return null;
		}
		try {
			return Long.valueOf(value);
		} catch (NumberFormatException e) {
			return null;
		}
	}

	/**
	 * 저장소에서 세어 Redis에 다시 넣는다.
	 *
	 * <p>읽은 시각이 없다는 것은 한 번도 안 읽었다는 뜻이지 읽을 것이 없다는 뜻이
	 * 아니다. 그때는 방의 처음부터 전부 센다. 예전에는 여기서 0을 돌려줬고, 그래서
	 * 새로 만든 방에 상대가 먼저 말을 걸면 배지에 아무것도 안 떴다.
	 */
	private long warmup(Long chatRoomId, Long memberId) {
		try {
			ChatRoomMember member = chatRoomMemberRepository
				.findByChatRoomIdAndMemberId(chatRoomId, memberId)
				.orElse(null);

			if (member == null) {
				log.warn("채팅방 멤버 없음: chatRoomId={}, memberId={}", chatRoomId, memberId);
				return 0L;
			}

			// 본인이 보낸 것은 어느 쪽이든 빼고 센다
			long actualCount = member.getLastReadAt() == null
				? chatMessageRepository
					.countByChatRoomIdAndIsDeletedFalseAndSenderIdNot(chatRoomId, memberId)
				: chatMessageRepository
					.countByChatRoomIdAndIsDeletedFalseAndCreatedAtAfterAndSenderIdNot(
						chatRoomId, member.getLastReadAt(), memberId);

			try {
				redis.opsForValue().set(getKey(chatRoomId, memberId),
					String.valueOf(actualCount), TTL_DAYS, TimeUnit.DAYS);
			} catch (Exception e) {
				log.error("Redis 캐싱 실패: {}", e.getMessage());
			}

			return actualCount;
		} catch (Exception e) {
			// 세는 데 실패해도 방 목록은 떠야 한다. 배지만 0으로 보인다.
			log.error("안읽음 집계 실패: chatRoomId={}, memberId={}, error={}",
				chatRoomId, memberId, e.getMessage(), e);
			return 0L;
		}
	}
}

File: backend/src/main/kotlin/com/joying/chat/service/UnreadCountService.kt
package com.joying.chat.service

import com.joying.chat.repository.ChatMessageRepository
import com.joying.chat.repository.ChatRoomMemberRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withContext
import org.slf4j.LoggerFactory
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.stereotype.Service
import java.util.concurrent.TimeUnit

/**
 * 안읽은 메시지 개수 관리 Service (Redis 캐싱)
 *
 * 전략:
 * - Redis: 캐시 (빠른 조회, 확장성)
 * - MongoDB: Source of Truth (정확성)
 * - 캐시 미스 시 MongoDB에서 계산 후 Redis에 저장
 */
@Service
class UnreadCountService(
    private val redis: RedisTemplate<String, String>,
    private val chatMessageRepository: ChatMessageRepository,
    private val chatRoomMemberRepository: ChatRoomMemberRepository
) {
    private val logger = LoggerFactory.getLogger(UnreadCountService::class.java)

    companion object {
        private const val UNREAD_KEY_PREFIX = "unread:"
        private const val TTL_DAYS = 7L
    }

    /**
     * Redis Key 생성
     */
    private fun getKey(chatRoomId: Long, memberId: Long): String {
        return "${UNREAD_KEY_PREFIX}${chatRoomId}:${memberId}"
    }

    /**
     * 안읽은 개수 증가 (메시지 전송 시)
     *
     * @param chatRoomId 채팅방 ID
     * @param memberId 회원 ID (안읽은 개수를 증가시킬 대상, 즉 수신자)
     */
    fun increment(chatRoomId: Long, memberId: Long) {
        try {
            val key = getKey(chatRoomId, memberId)
            redis.opsForValue().increment(key)
            redis.expire(key, TTL_DAYS, TimeUnit.DAYS)

            logger.debug("Redis 안읽은 개수 증가: chatRoomId={}, memberId={}", chatRoomId, memberId)
        } catch (e: Exception) {
            logger.error("Redis increment 실패 (MongoDB에는 저장됨, 재접속 시 복구 가능): ${e.message}")
        }
    }

    /**
     * 안읽은 개수 초기화 (읽음 처리 시)
     *
     * @param chatRoomId 채팅방 ID
     * @param memberId 회원 ID
     */
    fun reset(chatRoomId: Long, memberId: Long) {
        try {
            val key = getKey(chatRoomId, memberId)
            redis.delete(key)

            logger.debug("Redis 안읽은 개수 초기화: chatRoomId={}, memberId={}", chatRoomId, memberId)
        } catch (e: Exception) {
            logger.error("Redis reset 실패: ${e.message}")
        }
    }

    /**
     * 안읽은 개수 조회 (단일 채팅방)
     *
     * Redis 우선, 캐시 미스 시 MongoDB에서 계산
     *
     * @param chatRoomId 채팅방 ID
     * @param memberId 회원 ID
     * @return 안읽은 메시지 개수
     */
    suspend fun get(chatRoomId: Long, memberId: Long): Long {
        val key = getKey(chatRoomId, memberId)

        // 1. Redis 조회 (캐시 히트)
        val cached = redis.opsForValue().get(key)
        if (cached != null) {
            logger.debug("Redis 캐시 히트: chatRoomId={}, memberId={}, count={}", chatRoomId, memberId, cached)
            return cached.toLong()
        }

        // 2. 캐시 미스 → MongoDB에서 계산 후 Redis에 저장
        logger.debug("Redis 캐시 미스: chatRoomId={}, memberId={} → MongoDB 조회", chatRoomId, memberId)
        return warmup(chatRoomId, memberId)
    }

    /**
     * 안읽은 개수 배치 조회 (여러 채팅방)
     *
     * 채팅방 목록 조회 시 사용
     * Redis Pipeline으로 한 번에 조회 후, 캐시 미스는 MongoDB에서 병렬 계산
     *
     * @param chatRoomIds 채팅방 ID 목록
     * @param memberId 회원 ID
     * @return 채팅방별 안읽은 개수 (chatRoomId -> count)
     */
    suspend fun getBatch(
        chatRoomIds: List<Long>,
        memberId: Long
    ): Map<Long, Long> {
        logger.info("[getBatch] 시작: chatRoomIds={}, memberId={}", chatRoomIds, memberId)

        if (chatRoomIds.isEmpty()) {
            logger.info("[getBatch] chatRoomIds가 비어있음 → emptyMap 반환")
            return emptyMap()
        }

        val keys = chatRoomIds.map { getKey(it, memberId) }
        logger.info("[getBatch] Redis keys 생성 완료: keys={}", keys)

        // Redis Pipeline으로 한 방에 조회
        logger.info("[getBatch] Redis multiGet 시작...")
        val cached = try {
            val result = redis.opsForValue().multiGet(keys) ?: emptyList()
            logger.info("[getBatch] Redis multiGet 성공: result={}", result)
            result
        } catch (e: Exception) {
            logger.error("[getBatch] Redis multiGet 실패: ${e.message}", e)
            emptyList()
        }

        val result = mutableMapOf<Long, Long>()
        val missedIds = mutableListOf<Long>()

        chatRoomIds.forEachIndexed { index, chatRoomId ->
            val count = cached.getOrNull(index)?.toLongOrNull()
            if (count != null) {
                result[chatRoomId] = count  // 캐시 히트
            } else {
                missedIds.add(chatRoomId)  // 캐시 미스
            }
        }

        logger.debug(
            "Redis 배치 조회: 총 {}개, 캐시 히트 {}개, 캐시 미스 {}개",
            chatRoomIds.size,
            result.size,
            missedIds.size
        )

        // 캐시 미스 → MongoDB에서 병렬 계산
        if (missedIds.isNotEmpty()) {
            coroutineScope {
                missedIds.map { chatRoomId ->
                    async {
                        chatRoomId to warmup(chatRoomId, memberId)
                    }
                }.forEach { deferred ->
                    val (chatRoomId, count) = deferred.await()
                    result[chatRoomId] = count
                }
            }
        }

        return result
    }

    /**
     * Cache Warming (MongoDB → Redis 동기화)
     *
     * MongoDB에서 실제 안읽은 개수를 계산하고 Redis에 저장
     *
     * @param chatRoomId 채팅방 ID
     * @param memberId 회원 ID
     * @return 안읽은 메시지 개수
     */
    private suspend fun warmup(chatRoomId: Long, memberId: Long): Long {
        logger.info("[warmup] 시작: chatRoomId={}, memberId={}", chatRoomId, memberId)

        try {
            val member = withContext(Dispatchers.IO) {
                chatRoomMemberRepository
                    .findByChatRoomIdAndMemberId(chatRoomId, memberId)
                    .orElse(null)
            }

            logger.info("[warmup] chatRoomMember 조회 완료: chatRoomId={}, member exists={}", chatRoomId, member != null)

            // 채팅방 멤버가 없으면 0
            if (member == null) {
                logger.warn("채팅방 멤버 없음: chatRoomId={}, memberId={}", chatRoomId, memberId)
                return 0L
            }

            // MongoDB에서 실제 안읽은 개수 계산
            logger.info("[warmup] MongoDB 카운트 쿼리 시작: chatRoomId={}, lastReadAt={}", chatRoomId, member.lastReadAt)
            val actualCount = if (member.lastReadAt != null) {
                withContext(Dispatchers.IO) {
                    chatMessageRepository.countByChatRoomIdAndIsDeletedFalseAndCreatedAtAfter(
                        chatRoomId,
                        member.lastReadAt!!
                    )
                }
            } else {
                logger.info("[warmup] lastReadAt이 null이므로 0 반환")
                0L
            }

            logger.info("[warmup] MongoDB 카운트 쿼리 완료: chatRoomId={}, count={}", chatRoomId, actualCount)

            // Redis에 캐싱 (7일 TTL)
            try {
                val key = getKey(chatRoomId, memberId)
                redis.opsForValue().set(key, actualCount.toString(), TTL_DAYS, TimeUnit.DAYS)

                logger.info(
                    "Cache warming 완료: chatRoomId={}, memberId={}, count={}",
                    chatRoomId,
                    memberId,
                    actualCount
                )
            } catch (e: Exception) {
                logger.error("Redis 캐싱 실패: ${e.message}")
            }

            return actualCount
        } catch (e: Exception) {
            logger.error("[warmup] 실패 - chatRoomId={}, memberId={}, error={}", chatRoomId, memberId, e.message, e)
            // MongoDB 조회 실패 시 기본값 0 반환 (서비스 중단 방지)
            return 0L
        }
    }
}